scripts/aws_replicate.py

```diff
@@ -0,0 +1,376 @@
+                bucket = s3.Bucket(bucket_name)
+                try:
+                    for file in bucket.objects.all():
+                        existed_keys.add(bucket_name + "/" + file.key)
+                except Exception:
+                    pass
+        return existed_keys
+
+    def prepare(self):
+        """
+        Read data file info from manifest and organize them into  groups.
+        Each group contains files should be copied to same bucket
+        The groups will be push to the queue
+        """
+        if self.manifest_file.startswith("s3://"):
+            submitting_files, _ = get_fileinfo_list_from_s3_manifest(self.manifest_file)
+        else:
+            submitting_files, _ = get_fileinfo_list_from_manifest(self.manifest_file)
+
+        file_grps = exec_files_grouping(submitting_files)
+        mutexLock.acquire()
+        for _, files in file_grps.iteritems():
+            chunk_size = self.global_config.get("chunk_size", 1)
+            idx = 0
+            while idx < len(files):
+                workQueue.put(files[idx : idx + chunk_size])
+                idx = idx + chunk_size
+        mutexLock.release()
+
+        for th in self.thread_list:
+            th.start()
+
+    def run(self,):
+        global TOTAL_PROCESSED_FILES
+
+        start = timeit.default_timer()
+        # Wait for queue to empty
+        while not workQueue.empty():
+            time.sleep(5)
+            current = timeit.default_timer()
+            # running_time = current - start
+            # if int(running_time) % 10== 0:
+            logger.info("=======================================================")
+            logger.info("Total files processed {}".format(TOTAL_PROCESSED_FILES))
+            logger.info("Times in second {}".format(current - start))
+
+        # Notify threads it's time to exit
+        global EXIT_FLAG
+        EXIT_FLAG = 1
+
+        # Wait for all threads to complete
+        for t in self.thread_list:
+            t.join()
+
+        logger.info("Done")
+
+
+def exec_aws_copy(
+    threadName, files, source_bucket, target_bucket, global_config, copied_keys
+):
+    """
+    Call AWS SLI to copy a chunk of  files from a bucket to another bucket.
+    Intergrity check: After each chunk copy, check the returned md5 hashes
+                        with the ones provided in manifest.
+    If not match, re-copy. Log all the success and failure cases
+
+    Args:
+        threadName(str): thread name 
+        files(list): a list of files which should be copied to the same bucket
+        source_bucket(str): source bucket
+        target_bucket(str): target bucket
+        global_config(dict): user config
+        copied_keys(set): set of objects already in target buckets
+    
+    Returns:
+        None
+    """
+
+    config_chunk_size = global_config.get("chunk_size", 1)
+
+    index = 0
+
+    while index < len(files):
+        base_cmd = 'aws s3 cp s3://{} s3://{} --recursive --exclude "*"'.format(
```

Comment:
is there a particular reason to do it with the CLI? it seems like the boto3 S3 client would be fine for this